### PR TITLE
OOP Frogger Game

### PR DESCRIPTION
**Pull request by madmaxWMFU**
### Object-Oriented JavaScript

[Demo](https://madmaxwmfu.github.io/frogger_game) |
[Code base](https://github.com/madmaxWMFU/frogger_game)

Please, review my frogger game.

### CODE REVIEW

File: submissions/madmaxWMFU/frogger-game/index.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <meta charset="UTF-8">
+    <title>Effective JavaScript: Frogger</title>
+    <link rel="stylesheet" href="css/style.css">
+</head>
+<body>
+	<h3>Score: <span class="user-score">0</span></h3>
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/index.html]: Please, remove redundant space.

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    "x": 200,
+    "y": 400,
+    "icon": "images/char-boy.png"
+};
+
+let score = document.querySelector(".user-score");
+
+class Enemy {
+    constructor(x, y, speed) {
+        this.x = x;
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: `Enemy` and `Player` have a common properties (`x, y, sprite`). In OOP better way save it in a common class. IMO `Character` class will be good solution.

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        this.y = y;
+        this.sprite = "images/enemy-bug.png";
+        this.speed = speed;
+        this.player = player;
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: `player` property looks like a global. You pass it to constructor, but not access them.

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    
+    update(dt) {
+        this.x += this.speed * dt;
+        if(this.x > 510) {
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: All numbers is a magic. We can define them in `const`.

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            this.x = -50;
+            this.speed = randomSpeed();
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: `randomSpead()` - is a method for `Enemy` class. So, we can define them in class. Also methods must be named with verb. `getRandomSpead()` looks like better.

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        if(player.x < this.x + 80 && player.x + 80 > this.x && player.y < this.y + 60 && 60 + player.y > this.y) {
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: :eyes: magic numbers )

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        this.x = x;
+        this.y = y;
+        this.sprite = icon;
+    }
+
+    update(dt) {
+
+    }
+
+    render() {
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: This is also common method with `Enemy` class.

**Comment by madmaxWMFU** [comment]
@o-msh thanks for your feedback!
I corrected all the shortcomings

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    }
+    
+    update(dt) {
+        this.x += this.speed * dt;
+        if(this.x > enemyStartValue.maxX) {
+            this.x = enemyStartValue.minX;
+            this.speed = this.getRandomSpeed();
+        }
+
+        if(player.x < this.x + coordsGameGround.conflictX && player.x + coordsGameGround.conflictX > this.x && player.y < this.y + coordsGameGround.conflictY && coordsGameGround.conflictY + player.y > this.y) {
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: IMO `update` function should not be deal with collision solution. Can we define other method for this? Use **separation of concerns principle**.

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+            player.x = playerStartValue.x;
+            player.y = playerStartValue.y;
+            minusScore();
+        }
+    }
+
+    render() {
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: This method already defined in parent class, and, if i don't mistaken - this is unnecessary.

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        super.render();
+    }
+
+    getRandomSpeed() {
+        return 100 + Math.floor(Math.random() * 222);
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: Better will be define `const` for all numbers. Maybe some `minSpead` and `maxSpead`?

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+        if(keyPress == 'up' && this.y > coordsGameGround.min) {
+            this.y -= playerStartValue.stepY;
+        }
+
+        if(keyPress == 'down' && this.y < coordsGameGround.max) {
+            this.y += playerStartValue.stepY;
+        }
+      
+        if(this.y <= coordsGameGround.min) {
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: Please, split `handleInput` function for other calculation. Can we move this check in `update` function ?

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        38: 'up',
+        39: 'right',
+        40: 'down'
+    };
+
+    player.handleInput(allowedKeys[e.keyCode]);
+});
+
+const player = new Player(playerStartValue.x, playerStartValue.y, playerStartValue.icon);
+const allEnemies = [63, 147, 230].map(y =>  new Enemy(enemyStartValue.x, y, enemyStartValue.icon, player));
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: At first sight this numbers are incomprehensible. Can we write this more pretty?

**Comment by madmaxWMFU** [comment]
I forgot that I don't need to call the function `render()`  again.
I corrected all your commits.

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        this.player = player;
+    }
+    
+    update(dt) {
+        this.x += this.speed * dt;
+        if(this.x > enemyStartValue.maxX) {
+            this.x = enemyStartValue.minX;
+            this.speed = this.getRandomSpeed();
+        }
+        this.resetPlayer();
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: `resetPlayer` means that we reset player position? But we must check for collision player with enemies. So, more clearly will be change name of method to `isCollision` or `checkCollision` or it's up to you.

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    }
+}
+
+class Player extends Character {
+    constructor(x, y, sprite) {
+        super(x, y, sprite);
+    }
+
+    update(dt) {
+
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: Why you need this empty method?

File: submissions/madmaxWMFU/frogger-game/js/app.js [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+        if(keyPress == 'up' && this.y > coordsGameGround.min) {
+            this.y -= playerStartValue.stepY;
+        }
+
+        if(keyPress == 'down' && this.y < coordsGameGround.max) {
+            this.y += playerStartValue.stepY;
+        }
+        
+        this.prizePoint();
```
Comment (o-msh) [inline submissions/madmaxWMFU/frogger-game/js/app.js]: 'handleInput` method means that we handle input key press. I think we can move `prizePoint` method to `update` function. And you ?

Comment (madmaxWMFU) [reply submissions/madmaxWMFU/frogger-game/js/app.js]: If I move this func in `handleInput` the account is growing rapidly

File: submissions/madmaxWMFU/frogger-game/js/app.js (hunk 14 above)
Comment (madmaxWMFU) [reply submissions/madmaxWMFU/frogger-game/js/app.js]: Without this feature, the game does not work.
This is func only for update player

**Pull request merged**